README.md
# Deep Dream Travel

Deep Dream Travel allows you to turn a single image into a video, diving through various layers of a neural network.

This effect is achieved by recursively feeding the image into the network and regularly switching the targeted layer.

## Demo

![Deep Dream Demo](https://drive.google.com/uc?export=view&id=1sHDkmz13IQlp-M2KxSlwI6htXaL04wnH)

### Higher resolution video
A higher resolution example can be found on [YouTube](https://www.youtube.com/watch?v=VjcBpVmRm9Y).

## Installation
Create and activate a new Anaconda environment
```
conda create --name DeepDreamTravel
conda activate DeepDreamTravel
```
Install required dependencies
```
pip install -r requirements.txt
conda install psutil
```
Install Caffe depending on your operating system
```
conda install caffe <-- Linux
conda install caffe -c willyd <-- Windows
```

## Usage
### Command line
```
python deepdreamtravel.py INPUT_FILE OUTPUT_VIDEO
```
or view all options
```
python deepdreamtravel.py -h
```

### Within an application
View `example.py` for an example of running DeepDreamTravel inside your application.
Available arguments are listed bellow.
```python
from deepdreamtravel import DeepDreamTravel

dreamer = DeepDreamTravel('protxt', 'caffemodel') # View example.py for an example
dreamer.generate(input_image="noise.jpg",
                 node_switch=10, # How often should layers switch
                 resize_coeff=0.05, # Zoom amount in %
                 show_iter=100, # The image will be shown every X iterations
                 offset=[0,0,0,0], # Left, Top, Right, Bottom zoom offset
                 temp_dir="tmp", # Temp directory to store frames
                 start_iter=0, # If generation was interrupted, enter the next iteration here
                 start_index=0, # Which layer index should be used first. Also used when interruption occured
                 start_offset=0, # In-case there are invalid layers, enter the offset given by terminal output when interruption happened.
                 max_iteration=False, # If false, all layers will be "explored". Else set maximum number of iterations
                 octaves=4, # Higher number leads to bigger ("better") visuals. Takes significantly more time to generate! Lower number in-case of errors with small images.
                 iter_n=10, # Number of iterations per octave.
                 octave_scale=1.4, # Resize amount per octave. Higher number leads to higher dream states.
                 output_video="DeepDreamTravel.avi", # Output location and name. Don't forget to end with .avi!
                 fps=30, # Frames per second
                 delete_temp=True, # Clean up generated images after video was generated?
                 max_memory=3 # Maximum amount of memory to use in GB
)
```
**Note:** The examples were generated with [Places205-GoogLeNet](http://places.csail.mit.edu/downloadCNN.html). You can find other interesting models [here](https://github.com/BVLC/caffe/wiki/Model-Zoo).
## Dependencies

* [PIL](https://pillow.readthedocs.io/en/stable/)
* [SciPy](https://pypi.org/project/scipy/)
* [Caffe](https://caffe.berkeleyvision.org/)
* [OpenCV](https://pypi.org/project/opencv-python/)

## Support
If you are having issues with the installation, [Nerdy Rodent](https://www.youtube.com/channel/UC4-5v-f-xKnbi1yaAuRSi_w) was kind enough to create a [tutorial](https://www.youtube.com/watch?v=f1HLevIo0Z8).

Additionally, [support, questions, and requests are provided here](https://hackcommunity.net/t/trippy-videos-with-deep-dream-travel)

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

## License
[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.1] - 2021-01-17
### Added
- Ability to execute from command line
- Requirements.txt

### Changed
- Fixed incorrect file location handling

deepdreamtravel/utility.py
import numpy as np
import PIL.Image
from pathlib import Path


# a couple of utility functions for converting to and from Caffe's input image layout
def preprocess(net, img):
    return np.float32(np.rollaxis(img, 2)[::-1]) - net.transformer.mean['data']


def deprocess(net, img):
    return np.dstack((img + net.transformer.mean['data'])[::-1])


def objective_L2(dst):
    dst.diff[:] = dst.data


def save_images(images_arr, last_image, output_dir="", fmt='jpeg'):
    '''
    Saving all images to disk
    '''
    print("Writing files to disk.")
    img_num = last_image - (len(images_arr) - 1)
    for image in images_arr:
        image = image / 255.0
        image = np.uint8(np.clip(image, 0, 1) * 255)
        PIL.Image.fromarray(image).save(Path(output_dir, str(img_num) + ".jpg"), fmt)
        img_num += 1

    return_image = images_arr[-1]
    return [return_image]

imagetovideo/imagetovideo.py
import cv2
import os.path
from pathlib import Path


class ImageToVideo(object):

    @staticmethod
    def convert(dir_in, out, prefix="", first_image=0, extension=".jpg", fps=30, delete_files=False):
        """
        Converts images to avi file. Images should be numbered from 0 in order they should appear in video.

        :string dir_in: Directory where image files are located
        :string out: Final name of file (with .avi extension)
        :string or list of strings prefix: what prefixes do files use. Ideally, they should be in order for best runtime.
        :int first_image: Number of the first image
        :string extension: Extensions of files (.jpg etc.)
        :int fps: Frames per second to use
        :bool delete_files: Delete input images?
        :return: True if files have been written. Currently not much error handling is done.
        """
        img_array = []  # Array to hold images.
        file_exists = True
        file_num = first_image    # Counter for files to be written

        # Handle if input is a list
        lst = False
        lst_index = -1
        if type(prefix) is list:
            lst = True
            lst_index = 0   # Hold the index of prefix.

        filename, lst_index = ImageToVideo.get_filename(dir_in, prefix, file_num, extension, lst_index=lst_index)
        if not filename:
            return False
        img = cv2.imread(str(filename))
        height, width, layers = img.shape
        size = (width, height)
        out = cv2.VideoWriter(out, cv2.VideoWriter_fourcc(*'DIVX'), fps, size)

        # Writing files
        while file_exists:
            filename, lst_index = ImageToVideo.get_filename(dir_in, prefix, file_num, extension, lst_index)

            # If file does not exist (run out of images to write)
            if not filename:
                file_exists = False
                break

            img = cv2.imread(str(filename))
            out.write(img)
            if delete_files:
                os.unlink(filename)

            # Change this to += 1
            file_num += 1

        out.release()
        return True

    @staticmethod
    def get_filename(dir_in, prefix, file_num, extension, lst_index=-1):
        """
        Returns a filename or False if file does not exist.
        See convert() for params
        :int lst_index: -1 if prefix is a string else it will check indexes until it finds existing file or exhausts list.
                        The lst_index should improve runtime if prefixes are in order. Else it will search the whole list.
        :return: path to file if file exists else False. In case of list of prefixes it also returns the index
        """
        # Finding file for a single prefix
        if lst_index == -1:
            if os.path.isfile(Path(dir_in, str(file_num) + extension)):
                return Path(dir_in, prefix + str(file_num) + extension), -1
            else:
                return False, False

        # Finding file for list of prefixes
        prefix_len = len(prefix)
        if os.path.isfile(Path(dir_in, prefix[lst_index] + str(file_num) + extension)):
            return Path(dir_in, prefix[lst_index] + str(file_num) + extension), lst_index
        # Checking next index (and if it is not out of bounds)
        elif lst_index < prefix_len and os.path.isfile(Path(dir_in, prefix[lst_index + 1] + prefix[lst_index] + str(file_num) + extension)):
            return Path(dir_in, prefix[lst_index + 1] + str(file_num) + extension), lst_index + 1

        # If none of the above work, go through the whole list
        for i in range(prefix_len):
            if os.path.isfile(Path(dir_in, prefix[i] + str(file_num) + extension)):
                return Path(dir_in + prefix[i] + str(file_num) + extension), i

        # If after all this nothing is found, return false
        return False, False

deepdreamtravel/deepdreamtravel.py
from io import BytesIO
import PIL.Image
import io
import os
import psutil
from google.protobuf import text_format
import scipy.ndimage as nd
import caffe
from .utility import *
from deepdreamtravel.imagetovideo import ImageToVideo
from random import randint
from pathlib import Path

class DeepDreamTravel:
    objectives = []

    def __init__(self, prototxt, caffemodel, model_path="./"):
        caffe.set_device(0)
        caffe.set_mode_gpu()

        net_fn = model_path + prototxt
        param_fn = model_path + caffemodel

        # Patching model to be able to compute gradients.
        # Note that you can also manually add "force_backward: true" line to "deploy.prototxt".
        model = caffe.io.caffe_pb2.NetParameter()
        text_format.Merge(open(net_fn).read(), model)
        model.force_backward = True
        open('tmp.prototxt', 'w').write(str(model))

        self.net = caffe.Classifier('tmp.prototxt', param_fn,
                               mean=np.float32([104.0, 116.0, 122.0]),  # ImageNet mean, training set dependent
                               channel_swap=(2, 1, 0))  # the reference model has channels in BGR order instead of RGB

        self.nodes = list(self.net._layer_names)
        self.nodes = self.nodes[1:-2]
        self.node_len = len(self.nodes)
        self.objectives.append(objective_L2)
        print("Loaded ", self.node_len, "nodes")

    def register_objective(self, func):
        """
        Register a new objective to be used by deep dream
        :function func:
        :return: True if successful else False
        """
        if not hasattr(func, '__call__'):
            return False

        self.objectives.append(func)
        return True

    def make_step(self, net, step_size=1.5, end='inception_4c/output',
                  jitter=32, clip=True, objective=objective_L2):
        '''Basic gradient ascent step.'''

        src = net.blobs['data']  # input image is stored in Net's 'data' blob
        dst = net.blobs[end]

        ox, oy = np.random.randint(-jitter, jitter + 1, 2)
        src.data[0] = np.roll(np.roll(src.data[0], ox, -1), oy, -2)  # apply jitter shift

        net.forward(end=end)
        objective(dst)  # specify the optimization objective
        net.backward(start=end)
        g = src.diff[0]
        # apply normalized ascent step to the input image
        src.data[:] += step_size / np.abs(g).mean() * g

        src.data[0] = np.roll(np.roll(src.data[0], -ox, -1), -oy, -2)  # unshift image

        if clip:
            bias = net.transformer.mean['data']
            src.data[:] = np.clip(src.data, -bias, 255 - bias)

    def list_nodes(self, display=True):
        """
        :bool display: Print out all nodes
        :return: list of nodes
        """
        if display:
            print("The following nodes are loaded:", self.nodes)
        return self.nodes

    def showarray(self, a, fmt='jpeg'):
        # clip the values to be between 0 and 255
        a = np.uint8(np.clip(a, 0, 255))
        f = BytesIO()
        PIL.Image.fromarray(a).save(f, fmt)
        return f.getvalue()

    def deepdream(self, net, base_img, iter_n=10, octave_n=4, octave_scale=1.4,
                     end='inception_4c/output', clip=True, **step_params):
        # prepare base images for all octaves
        octaves = [preprocess(net, base_img)]
        for i in range(octave_n - 1):
            octaves.append(nd.zoom(octaves[-1], (1, 1.0 / octave_scale, 1.0 / octave_scale), order=1))

        src = net.blobs['data']
        detail = np.zeros_like(octaves[-1])  # allocate image for network-produced details
        for octave, octave_base in enumerate(octaves[::-1]):
            h, w = octave_base.shape[-2:]
            if octave > 0:
                # upscale details from the previous octave
                h1, w1 = detail.shape[-2:]
                detail = nd.zoom(detail, (1, 1.0 * h / h1, 1.0 * w / w1), order=1)

            src.reshape(1, 3, h, w)  # resize the network's input image size
            src.data[0] = octave_base + detail
            for i in range(iter_n):
                self.make_step(net, end=end, clip=clip, **step_params)

                # visualization
                vis = deprocess(net, src.data[0])
                if not clip:  # adjust image contrast if clipping is disabled
                    vis = vis * (255.0 / np.percentile(vis, 99.98))
                ret = self.showarray(vis)

            # extract details produced on the current octave
            detail = src.data[0] - octave_base
        # returning the resulting image
        return ret

    def generate(self,
                 input_image="noise.jpg",
                 node_switch=10,
                 resize_coeff=0.05,
                 show_iter=100,
                 offset=[0,0,0,0],
                 temp_dir="tmp",
                 start_iter=0,
                 start_index=0,
                 start_offset=0,
                 max_iteration=False,
                 octaves=4,
                 iter_n=10,
                 octave_scale=1.4,
                 output_video="DeepDreamTravel.avi",
                 fps=30,
                 delete_temp=True,
                 max_memory=3
                 ):
        # Loading the image
        img0 = PIL.Image.open(input_image)

        w, h = img0.size
        img0 = np.float32(img0)

        images_arr = [img0]  # Storing all images in an array

        node_switch = node_switch  # How often should nodes be switched

        image_resize_coeff = resize_coeff
        max_memory = max_memory  # in GB
        show_image = show_iter  # Show image every X iterations
        offset = offset  # Left, Top, Right, Bottom

        directory = Path(temp_dir)
        if not os.path.exists(directory):
            os.makedirs(directory)

        # Gettin process ID to monitor memory
        process = psutil.Process(os.getpid())

        iteration = start_iter
        index = start_index  # Starting at first node (start at 41)
        index_offset = start_offset  # To display true index from the original list add offset

        # If Max iteration is not set, the video will include all layers
        max_iter = max_iteration if max_iteration is not False else self.node_len * node_switch
        if max_iteration is False:
            print("Max iteration not set. Setting to", max_iter)

        objective_index_max = len(self.objectives) # Number of registered objectives
        while iteration < max_iter:
            node = self.nodes[index]
            # Getting a randomg objective
            objective = self.objectives[randint(0, objective_index_max - 1)]

            if (process.memory_info().rss * 10 ** -9) >= max_memory:
                images_arr = save_images(images_arr, iteration, output_dir=directory)

            try:
                for octave in range(1, octaves + 1):
                    for _ in range(int(node_switch / octaves)):
                        print(node, "iteration", iteration, "index", index, "offset", index_offset)
                        img0 = self.deepdream(self.net, images_arr[-1], iter_n=iter_n, octave_n=octave,
                                              octave_scale=octave_scale, end=node, objective=objective)
                        img0 = PIL.Image.open(io.BytesIO(img0))

                        if not iteration % show_image:
                            img0.show()
                        # Cropping image
                        img0 = nd.affine_transform(img0, [1 - image_resize_coeff + offset[0], 1 - image_resize_coeff + offset[1], 1],
                                                   [h * image_resize_coeff / 2 - offset[2], w * image_resize_coeff / 2 - offset[3], 0], order=1)
                        # Resizing image
                        img1 = np.float32(img0)
                        images_arr.append(img1)
                        iteration += 1
            except (ValueError, KeyError):
                print("Invalid node")
                index_offset += 1
                del self.nodes[index]
                self.node_len = len(self.nodes)

                # Adjusting max iteration
                if max_iteration is False:
                    max_iter = self.node_len * node_switch
                    print("Adjusted max iteration to", max_iter)
                continue
            index += 1
            if index >= self.node_len:
                print("Too large node.")
                index = 0
        save_images(images_arr, iteration, output_dir=directory)

        # Generating video
        print("Generating video...")
        # Making sure the output_video ends with .avi
        if Path(output_video).suffix != '.avi':
            output_video += '.avi'

        file_generated = ImageToVideo.convert(temp_dir, output_video, fps=fps, delete_files=delete_temp)
        if file_generated:
            print("Video generated to", output_video)
        else:
            print("Error File not generated.")